File: experiments/test_issue25.pas
program test_issue25;

{$mode objfpc}{$H+}

uses
  SysUtils, Math;

type
  TVector3D = record
    x, y, z: Double;
  end;

  TMatrix4D = array[0..3, 0..3] of Double;

function IdentityMatrix: TMatrix4D;
var
  i, j: Integer;
begin
  for i := 0 to 3 do
    for j := 0 to 3 do
      if i = j then
        Result[i, j] := 1.0
      else
        Result[i, j] := 0.0;
end;

function CreateRotationMatrixZ(angleRad: Double): TMatrix4D;
var
  c, s: Double;
begin
  Result := IdentityMatrix;
  c := Cos(angleRad);
  s := Sin(angleRad);
  Result[0, 0] := c;
  Result[0, 1] := -s;
  Result[1, 0] := s;
  Result[1, 1] := c;
end;

function CreateTranslationMatrix(dx, dy, dz: Double): TMatrix4D;
begin
  Result := IdentityMatrix;
  Result[3, 0] := dx;
  Result[3, 1] := dy;
  Result[3, 2] := dz;
end;

function MultiplyMatrix(const A, B: TMatrix4D): TMatrix4D;
var
  i, j, k: Integer;
begin
  for i := 0 to 3 do
    for j := 0 to 3 do begin
      Result[i, j] := 0.0;
      for k := 0 to 3 do
        Result[i, j] := Result[i, j] + A[i, k] * B[k, j];
    end;
end;

function TransformPoint(const P: TVector3D; const M: TMatrix4D): TVector3D;
begin
  Result.x := P.x * M[0, 0] + P.y * M[1, 0] + P.z * M[2, 0] + M[3, 0];
  Result.y := P.x * M[0, 1] + P.y * M[1, 1] + P.z * M[2, 1] + M[3, 1];
  Result.z := P.x * M[0, 2] + P.y * M[1, 2] + P.z * M[2, 2] + M[3, 2];
end;

function TransformAngle(angle, rotationAngle: Double): Double;
begin
  Result := angle + rotationAngle;
  while Result >= 360.0 do
    Result := Result - 360.0;
  while Result < 0.0 do
    Result := Result + 360.0;
end;

var
  arcCenter, rotCenter, newCenter: TVector3D;
  startAngle, endAngle, rotAngle: Double;
  radius: Double;
  M, T1, R, T2: TMatrix4D;
  newStartAngle, newEndAngle: Double;

begin
  WriteLn('=== Test Issue #25: Arc Rotation ===');
  WriteLn;

  arcCenter.x := 2.0;
  arcCenter.y := 5.0;
  arcCenter.z := 0.0;
  radius := 10.0;
  startAngle := 8.0;
  endAngle := 94.0;

  WriteLn('Original Arc:');
  WriteLn('  Center: (', arcCenter.x:0:4, ', ', arcCenter.y:0:4, ', ', arcCenter.z:0:4, ')');
  WriteLn('  Radius: ', radius:0:4);
  WriteLn('  Start Angle: ', startAngle:0:4, '°');
  WriteLn('  End Angle: ', endAngle:0:4, '°');
  WriteLn;

  rotCenter.x := 1.0;
  rotCenter.y := 1.0;
  rotCenter.z := 1.0;
  rotAngle := 25.0;

  WriteLn('Rotation:');
  WriteLn('  Around point: (', rotCenter.x:0:4, ', ', rotCenter.y:0:4, ', ', rotCenter.z:0:4, ')');
  WriteLn('  Angle: ', rotAngle:0:4, '°');
  WriteLn;

  T1 := CreateTranslationMatrix(-rotCenter.x, -rotCenter.y, -rotCenter.z);
  R := CreateRotationMatrixZ(DegToRad(rotAngle));
  T2 := CreateTranslationMatrix(rotCenter.x, rotCenter.y, rotCenter.z);

  M := MultiplyMatrix(T1, R);
  M := MultiplyMatrix(M, T2);

  newCenter := TransformPoint(arcCenter, M);
  newStartAngle := TransformAngle(startAngle, rotAngle);
  newEndAngle := TransformAngle(endAngle, rotAngle);

  WriteLn('Transformed Arc (Simple Calculation):');
  WriteLn('  Center: (', newCenter.x:0:4, ', ', newCenter.y:0:4, ', ', newCenter.z:0:4, ')');
  WriteLn('  Radius: ', radius:0:4);
  WriteLn('  Start Angle: ', newStartAngle:0:4, '°');
  WriteLn('  End Angle: ', newEndAngle:0:4, '°');
  WriteLn;

  WriteLn('Expected (from issue #25):');
  WriteLn('  Center: (-0.2158, 5.0478, 0.0000)');
  WriteLn('  Radius: 10.0000');
  WriteLn('  Start Angle: 33.0000°');
  WriteLn('  End Angle: 119.0000°');
  WriteLn;

  ReadLn;
end.

File: experiments/test_issue25.py
#!/usr/bin/env python3
"""
Test script to verify arc rotation for issue #25
"""

import math
import numpy as np

def create_rotation_matrix_z(angle_deg):
    """Create a 4x4 rotation matrix for rotation around Z axis"""
    angle_rad = math.radians(angle_deg)
    c = math.cos(angle_rad)
    s = math.sin(angle_rad)
    return np.array([
        [c, -s, 0, 0],
        [s,  c, 0, 0],
        [0,  0, 1, 0],
        [0,  0, 0, 1]
    ])

def create_translation_matrix(dx, dy, dz):
    """Create a 4x4 translation matrix"""
    return np.array([
        [1, 0, 0, dx],
        [0, 1, 0, dy],
        [0, 0, 1, dz],
        [0, 0, 0,  1]
    ])

def transform_point(point, matrix):
    """Transform a 3D point using a 4x4 matrix"""
    p = np.array([point[0], point[1], point[2], 1])
    result = matrix @ p
    return result[:3]

def main():
    print("=== Test Issue #25: Arc Rotation ===\n")

    # Original arc parameters
    arc_center = np.array([2.0, 5.0, 0.0])
    radius = 10.0
    start_angle = 8.0
    end_angle = 94.0

    print("Original Arc:")
    print(f"  Center: ({arc_center[0]:.4f}, {arc_center[1]:.4f}, {arc_center[2]:.4f})")
    print(f"  Radius: {radius:.4f}")
    print(f"  Start Angle: {start_angle:.4f}°")
    print(f"  End Angle: {end_angle:.4f}°")
    print()

    # Rotation parameters
    rot_center = np.array([1.0, 1.0, 1.0])
    rot_angle = 25.0

    print("Rotation:")
    print(f"  Around point: ({rot_center[0]:.4f}, {rot_center[1]:.4f}, {rot_center[2]:.4f})")
    print(f"  Angle: {rot_angle:.4f}°")
    print()

    # Create transformation matrix: T^-1 * R * T
    # Move rotation center to origin, rotate, move back
    T1 = create_translation_matrix(-rot_center[0], -rot_center[1], -rot_center[2])
    R = create_rotation_matrix_z(rot_angle)
    T2 = create_translation_matrix(rot_center[0], rot_center[1], rot_center[2])

    # Combine: first T1, then R, then T2
    M = T2 @ R @ T1

    # Transform the arc center
    new_center = transform_point(arc_center, M)

    # Transform angles (simple addition for Z-axis rotation)
    new_start_angle = (start_angle + rot_angle) % 360
    new_end_angle = (end_angle + rot_angle) % 360

    print("Transformed Arc (Simple Z-axis Rotation):")
    print(f"  Center: ({new_center[0]:.4f}, {new_center[1]:.4f}, {new_center[2]:.4f})")
    print(f"  Radius: {radius:.4f}")
    print(f"  Start Angle: {new_start_angle:.4f}°")
    print(f"  End Angle: {new_end_angle:.4f}°")
    print()

    print("Expected (from issue #25):")
    print("  Center: (-0.2158, 5.0478, 0.0000)")
    print("  Radius: 10.0000")
    print("  Start Angle: 33.0000°")
    print("  End Angle: 119.0000°")
    print()

    # Check if results match
    expected_center = np.array([-0.2158, 5.0478, 0.0])
    expected_start = 33.0
    expected_end = 119.0

    center_error = np.linalg.norm(new_center - expected_center)
    start_error = abs(new_start_angle - expected_start)
    end_error = abs(new_end_angle - expected_end)

    print("Errors:")
    print(f"  Center error: {center_error:.6f}")
    print(f"  Start angle error: {start_error:.6f}°")
    print(f"  End angle error: {end_error:.6f}°")
    print()

    # Note: There might be a sign error in the expected center from the issue
    print("Note: The X coordinate sign differs.")
    print(f"  Our calculation: X={new_center[0]:.4f}")
    print(f"  Issue expected: X={expected_center[0]:.4f}")
    print("  This could be due to rotation direction (CW vs CCW)")
    print()

    if center_error < 0.001 and start_error < 0.1 and end_error < 0.1:
        print("✓ Test PASSED - Results match expected values")
    elif abs(new_center[0] + expected_center[0]) < 0.001 and abs(new_center[1] - expected_center[1]) < 0.001:
        print("⚠ Sign difference in X coordinate - rotation direction may be opposite")
    else:
        print("✗ Test FAILED - Results do not match")

    # Now let's also check what the actual points on the arc would be
    print("\n=== Verifying Arc Points ===")

    # Calculate start and end points on original arc
    start_rad = math.radians(start_angle)
    end_rad = math.radians(end_angle)

    orig_start_point = arc_center + np.array([
        radius * math.cos(start_rad),
        radius * math.sin(start_rad),
        0.0
    ])
    orig_end_point = arc_center + np.array([
        radius * math.cos(end_rad),
        radius * math.sin(end_rad),
        0.0
    ])

    print(f"Original start point: ({orig_start_point[0]:.4f}, {orig_start_point[1]:.4f}, {orig_start_point[2]:.4f})")
    print(f"Original end point: ({orig_end_point[0]:.4f}, {orig_end_point[1]:.4f}, {orig_end_point[2]:.4f})")

    # Transform these points
    new_start_point = transform_point(orig_start_point, M)
    new_end_point = transform_point(orig_end_point, M)

    print(f"Transformed start point: ({new_start_point[0]:.4f}, {new_start_point[1]:.4f}, {new_start_point[2]:.4f})")
    print(f"Transformed end point: ({new_end_point[0]:.4f}, {new_end_point[1]:.4f}, {new_end_point[2]:.4f})")

    # Calculate angles relative to new center
    relative_start = new_start_point - new_center
    relative_end = new_end_point - new_center

    calc_start_angle = math.degrees(math.atan2(relative_start[1], relative_start[0]))
    calc_end_angle = math.degrees(math.atan2(relative_end[1], relative_end[0]))

    if calc_start_angle < 0:
        calc_start_angle += 360
    if calc_end_angle < 0:
        calc_end_angle += 360

    print(f"Calculated start angle from point: {calc_start_angle:.4f}°")
    print(f"Calculated end angle from point: {calc_end_angle:.4f}°")

if __name__ == "__main__":
    main()
